src/relations/vault.py

```diff
@@ -0,0 +1,112 @@
+
+VAULT_NONCE_SECRET_LABEL = "nonce"  # nosec
+VAULT_CERT_PATH = "/vault/cert.pem"
+
+
+class VaultRelation(framework.Object):
+    """Client for vault relation."""
+
+    def __init__(self, charm):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+        """
+        super().__init__(charm, "vault")
+        self.charm = charm
+
+        charm.framework.observe(charm.vault.on.connected, self._on_vault_connected)
+        charm.framework.observe(charm.vault.on.ready, self._on_vault_ready)
+        charm.framework.observe(charm.vault.on.gone_away, self._on_vault_gone_away)
+
+    @log_event_handler(logger)
+    def _on_vault_connected(self, event: vault_kv.VaultKvConnectedEvent):
+        """Handle Vault connected event.
+
+        Args:
+            event: The event triggered when the Vault connection is created.
+        """
+        relation = self.charm.model.get_relation(event.relation_name, event.relation_id)
+        egress_subnet = str(self.charm.model.get_binding(relation).network.interfaces[0].subnet)
+        self.charm.vault.request_credentials(relation, egress_subnet, self.get_vault_nonce())
+
+    @log_event_handler(logger)
+    def _on_vault_ready(self, event: vault_kv.VaultKvReadyEvent):
+        """Handle Vault ready event.
+
+        Args:
+            event: The event triggered when the Vault connection is ready.
+        """
+        self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_vault_gone_away(self, event: vault_kv.VaultKvGoneAwayEvent):
+        """Handle Vault removed event.
+
+        Args:
+            event: The event triggered when the Vault connection is removed.
+        """
+        self.charm._update(event)
+
+    def get_vault_nonce(self):
+        """Retrieve the Vault nonce.
+
+        Returns:
+            The nonce retrieved from the secret storage.
+
+        Raises:
+            ModelError: If secret is not found.
+        """
+        try:
+            secret = self.charm.model.get_secret(label=VAULT_NONCE_SECRET_LABEL)
+            nonce = secret.get_content(refresh=True)["nonce"]
+            return nonce
+        except ModelError as e:
+            logger.debug(f"Secret {VAULT_NONCE_SECRET_LABEL} not found: {e}")
+            raise ModelError from e
+
+    def _get_vault_config(self):
+        """Retrieve Vault configuration details.
+
+        Returns:
+            A dictionary containing Vault configuration details if it exists
+        """
+        relation = self.charm.model.get_relation("vault")
+        if relation is None:
+            logger.debug("No vault relation found")
+            return None
+        vault_url = self.charm.vault.get_vault_url(relation)
+        ca_certificate = self.charm.vault.get_ca_certificate(relation)
+        mount = self.charm.vault.get_mount(relation)
+        unit_credentials = self.charm.vault.get_unit_credentials(relation)
+        if not unit_credentials:
```

Comment:
If there is a vault relation but no unit credentials do we need to put this in a blocked status? Or happy for it to default in this case to not using Vault?

Reply:
Good point :+1: Added logic to send it to a blocked state instead.